standards/SPIFFE_Federation.md

```diff
@@ -0,0 +1,316 @@
+
+Deprecated keys SHOULD be removed from the trust bundle after a trust domain no longer has any active, valid SVIDs issued from those keys. Failure to follow these recommendations as keys are added to and removed from the bundle could result in transient cross-domain authentication failures. Requirements for updating the trust bundle do not apply to keys used to issue SVIDs for internal use only.
+
+Clients SHOULD periodically poll the endpoint for updates because the contents are expected to [change over time][4] - key validity periods on the order of weeks or even days is commonplace. Clients SHOULD poll at a frequency equal to the value of the bundle’s `spiffe_refresh_hint`, in seconds. If not set, a reasonably low default value should apply - five minutes is recommended.
+
+### 4.2. Managing Fetched Bundles
+Clients of the bundle endpoint MUST store the latest SPIFFE bundle each time it is retrieved. The sequence number field of the trust bundle SHOULD be used when comparing the freshness or ordering of two trust bundles. If the trust bundle omits the sequence number, operators SHOULD consider the most recently retrieved bundle to be up-to-date.
+
+Operators MAY locally update the SPIFFE bundle of a foreign trust domain at any time. In this case, the locally updated version of the bundle is considered the latest until replaced by a subsequent refresh.
+
+Bundle contents from different trust domains MUST NOT be merged into a single, larger bundle. Doing so would enable one trust domain to forge identities belonging to another trust domain in the eyes of the validator with the unified bundle. As such, it is very important to ensure that bundles received from foreign trust domains be kept distinct, and clearly reflect the trust domain name to which they belong. Please see the [Security Considerations](#7-security-considerations) section for more information.
+
+### 4.3. Endpoint Address Stability
+Once foreign trust domains begin relying on a specific endpoint URL, it is a delicate and error-prone procedure to migrate all clients of the endpoint to a replacement endpoint URL. Therefore, the safest course of action is to prefer a stable endpoint URL.
+
+## 5. Serving and Consuming a SPIFFE Bundle Endpoint
+This specification defines two supported profiles for SPIFFE bundle endpoint servers, both of which are based on HTTPS. One relies on the use of Web PKI to authenticate the endpoint and the other leverages SPIFFE authentication. SPIFFE bundle endpoint clients MUST support both of these profiles while SPIFFE bundle endpoint servers MUST support at least one.
```

Comment:
Can we replace both "MUST"s in the last sentence with "SHOULD"s?

A SPIFFE deployment could reasonably use an authentication mechanism that isn't TLS-based to fetch a SPIFFE bundle. For this to occur, we'd also need to relax the condition that a SPIFFE bundle endpoint MUST be an HTTPS server and say instead that it SHOULD be an HTTPS server.

Reply:
cc: @justinburke

Reply:
> A SPIFFE deployment could reasonably use an authentication mechanism that isn't TLS-based to fetch a SPIFFE bundle.

I don't disagree, but this spec as written is based on TLS / HTTPS.  A different authentication mechanism would be a new profile, distinct from the ones defined in the spec.  We want it to be the case that if the client software complies with the spec and the server software complies with the spec, then it will work.  If we relax this to SHOULD, then you could have server software that doesn't use TLS and a client that insists on it, and interop will fail.

Reply:
> If we relax this to SHOULD, then you could have server software that doesn't use TLS and a client that insists on it, and interop will fail.

+1. The spec is about interop, and conformant software should work together. I don't think that means you can't write software that doesn't conform! Of course, you are free to write the software of your choosing ... however, if you write software that only supports an unspecified profile, then the spec has not accomplished much and you cannot reliably advertise conformance to your consumers and expect that to work. For example, even in a big corporation, if you say that you have a conformant SPIFFE federation thing but it uses a custom profile, clients or customers can't pick up off-the-shelf software and expect that to work.

Reply:
Thanks both for your replies. That's totally fair - I understand that you want any client/server pair using software that satisfies the spec to be able to fetch each other's trust bundle. 

I was coming at this question from the angle of: instead of having the control plane or trusted third party push the endpoint parameters, just have it push the SPIFFE bundle instead, and see if reading the bundle from disk could be made into a new endpoint profile. I'll file an issue to discuss further, since I don't want to clutter the PR. :)